File: server/src/main/java/org/jetbrains/bsp/bazel/server/bsp/utils/SourceRootGuesser.java
package org.jetbrains.bsp.bazel.server.bsp.utils;

import java.net.URI;
import java.nio.file.FileSystem;
import java.nio.file.FileSystems;
import java.nio.file.Path;
import java.nio.file.PathMatcher;
import java.nio.file.Paths;
import java.util.Objects;
import java.util.Optional;
import java.util.stream.Stream;

public class SourceRootGuesser {

  public static Path getSourcesRoot(Path sourcePath) {
    FileSystem fs = FileSystems.getDefault();
    PathMatcher sourceRootPattern =
        fs.getPathMatcher(
            "glob:**/"
                + "{main,test,tests,src,3rdparty,3rd_party,thirdparty,third_party}/"
                + "{*resources,scala,java,kotlin,jvm,proto,python,protobuf,py}");
    PathMatcher defaultTestRootPattern = fs.getPathMatcher("glob:**/{test,tests}");

    Optional<Path> sourceRootGuess =
        Stream.of(sourceRootPattern, defaultTestRootPattern)
            .map(pattern -> approximateSourceRoot(sourcePath, pattern))
            .filter(Objects::nonNull)
            .findFirst();

    return sourceRootGuess.orElse(sourcePath.getParent()).toAbsolutePath();
  }

  public static String getSourcesRoot(URI sourceUri) {
    return getSourcesRoot(Paths.get(sourceUri)).toString();
  }

  private static Path approximateSourceRoot(Path dir, PathMatcher matcher) {
    Path guess = dir;
    while (guess != null) {
      if (matcher.matches(guess)) {
        return guess;
      } else {
        guess = guess.getParent();
      }
    }
    return null;
  }
}

File: server/src/main/java/org/jetbrains/bsp/bazel/server/bsp/utils/SourceRootGuesser.kt
package org.jetbrains.bsp.bazel.server.bsp.utils

import java.net.URI
import java.nio.file.FileSystems
import java.nio.file.Path
import java.nio.file.PathMatcher
import java.nio.file.Paths

object SourceRootGuesser {
    fun getSourcesRoot(sourcePath: Path): Path {
        val fs = FileSystems.getDefault()
        val sourceRootPattern = fs.getPathMatcher(
            "glob:**/"
                    + "{main,test,tests,src,3rdparty,3rd_party,thirdparty,third_party}/"
                    + "{*resources,scala,java,kotlin,jvm,proto,python,protobuf,py}"
        )
        val defaultTestRootPattern = fs.getPathMatcher("glob:**/{test,tests}")
        val sourceRootGuess = sequenceOf(sourceRootPattern, defaultTestRootPattern)
            .mapNotNull { pattern: PathMatcher -> approximateSourceRoot(sourcePath, pattern) }
            .firstOrNull()
        return (sourceRootGuess ?: sourcePath.parent).toAbsolutePath()
    }

    fun getSourcesRoot(sourceUri: URI): String =
        getSourcesRoot(Paths.get(sourceUri)).toString()

    private fun approximateSourceRoot(dir: Path, matcher: PathMatcher): Path? {
        var guess: Path? = dir
        while (guess != null) {
            guess = if (matcher.matches(guess)) {
                return guess
            } else {
                guess.parent
            }
        }
        return null
    }
}
